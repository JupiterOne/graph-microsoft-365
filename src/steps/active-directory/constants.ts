// Step IDs
export const STEP_ACCOUNT = 'account';
export const STEP_GROUPS = 'groups';
export const STEP_GROUP_MEMBERS = 'group-members';
export const STEP_USERS = 'users';

// Graph objects
export const ACCOUNT_ENTITY_TYPE = 'microsoft_365_account';
export const ACCOUNT_ENTITY_CLASS = 'Account';

export const GROUP_ENTITY_TYPE = 'microsoft_365_user_group';
export const GROUP_ENTITY_CLASS = 'UserGroup';

export const USER_ENTITY_TYPE = 'microsoft_365_user';
export const USER_ENTITY_CLASS = 'User';

/**
 * The entity type used for members of groups which are not one of the ingested
 * directory objects.
 */
export const GROUP_MEMBER_ENTITY_TYPE = 'microsoft_365_group_member';

/**
 * The entity class used for members of groups which are not one of the ingested
 * directory objects.
 */
export const GROUP_MEMBER_ENTITY_CLASS = 'User';

export const ACCOUNT_GROUP_RELATIONSHIP_TYPE =
  'microsoft_365_account_has_group';
export const ACCOUNT_USER_RELATIONSHIP_TYPE = 'microsoft_365_account_has_user';
export const GROUP_MEMBER_RELATIONSHIP_TYPE = 'microsoft_365_group_has_member';
