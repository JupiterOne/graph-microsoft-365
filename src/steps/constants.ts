export const IngestionSources = {
  GROUPS: 'groups',
  USERS: 'users',
  APPLICATIONS: 'applications',
  DEVICES: 'devices',
  COMPLIANCE_POLICIES: 'compliance_policies',
  DEVICE_CONFIGURATIONS: 'device_configurations',
};
