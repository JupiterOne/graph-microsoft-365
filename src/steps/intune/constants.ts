import {
  RelationshipClass,
  StepEntityMetadata,
  StepRelationshipMetadata,
} from '@jupiterone/integration-sdk-core';
import { entities as activeDirectoryEntities } from '../active-directory';

export const steps: { [k: string]: string } = {
  FETCH_DEVICES: 'managed-devices',
  FETCH_DEVICE_CONFIGURATIONS_AND_FINDINGS:
    'device-configurations-and-findings',
  FETCH_MANAGED_APPLICATIONS: 'managed-applications',
  FETCH_DETECTED_APPLICATIONS: 'detected-applications',
};

export const entities: { [k: string]: StepEntityMetadata } = {
  DEVICE: {
    resourceName: 'Managed Device',
    _type: 'intune_managed_device',
    _class: ['Device', 'Host'],
  },
  DEVICE_CONFIGURATION: {
    resourceName: 'Device Configuration',
    _type: 'intune_device_configuration',
    _class: 'Configuration',
  },
  NONCOMPLIANCE_FINDING: {
    resourceName: 'Noncompliance Finding',
    _type: 'intune_noncompliance_finding',
    _class: 'Finding',
  },
  MANAGED_APPLICATION: {
    resourceName: 'Managed Application',
    _type: 'intune_managed_application',
    _class: 'Application',
  },
  DETECTED_APPLICATION: {
    resourceName: 'Detected Application',
    _type: 'intune_detected_application',
    _class: 'Application',
  },
};

export const relationships: { [k: string]: StepRelationshipMetadata } = {
  USER_HAS_DEVICE: {
    _type: 'azure_user_has_intune_managed_device',
    sourceType: activeDirectoryEntities.USER._type,
    _class: RelationshipClass.HAS,
    targetType: entities.DEVICE._type,
  },
  DEVICE_USES_DEVICE_CONFIGURATION: {
    _type: 'intune_managed_device_uses_device_configuration',
    sourceType: entities.DEVICE._type,
    _class: RelationshipClass.USES,
    targetType: entities.DEVICE_CONFIGURATION._type,
  },
  DEVICE_CONFIGURATION_IDENTIFIED_NONCOMPLIANCE_FINDING: {
    _type: 'intune_device_configuration_identified_noncompliance_finding',
    sourceType: entities.DEVICE_CONFIGURATION._type,
    _class: RelationshipClass.IDENTIFIED,
    targetType: entities.NONCOMPLIANCE_FINDING._type,
  },
  DEVICE_HAS_NONCOMPLIANCE_FINDING: {
    _type: 'intune_managed_device_has_noncompliance_finding',
    sourceType: entities.DEVICE._type,
    _class: RelationshipClass.HAS,
    targetType: entities.NONCOMPLIANCE_FINDING._type,
  },
  DEVICE_ASSIGNED_MANAGED_APPLICATION: {
    _type: 'intune_managed_device_assigned_application',
    sourceType: entities.DEVICE._type,
    _class: RelationshipClass.ASSIGNED,
    targetType: entities.MANAGED_APPLICATION._type,
  },
  DEVICE_HAS_DETECTED_APPLICATION: {
    _type: 'intune_managed_device_has_detected_application',
    sourceType: entities.DEVICE._type,
    _class: RelationshipClass.HAS,
    targetType: entities.DETECTED_APPLICATION._type,
  },
};
