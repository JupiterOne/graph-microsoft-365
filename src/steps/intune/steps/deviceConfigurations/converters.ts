import { createIntegrationEntity } from '@jupiterone/integration-sdk-core';
import { DeviceConfiguration } from '@microsoft/microsoft-graph-types-beta';
import { entities } from '../../constants';
import { last } from 'lodash';
import { DeviceConfigurationEntity } from '../../types';

/**
 * There are several different types of device configuration based on your OS and what you are configuring.
 * So many that it does not make sense to ingest each individually until that granularity is desired,
 * or necessary. Because of that, this method ingests all device configurations the same way, and if
 * more granularity is desired, individual configuration settings can be viewed in the raw data.
 *
 * https://docs.microsoft.com/en-us/graph/api/resources/intune-shared-deviceconfiguration?view=graph-rest-beta
 */
export function createDeviceConfigurationEntity(
  deviceConfiguration: DeviceConfiguration,
): DeviceConfigurationEntity {
  return createIntegrationEntity({
    entityData: {
      source: deviceConfiguration,
      assign: {
        _class: entities.DEVICE_CONFIGURATION._class,
        _type: entities.DEVICE_CONFIGURATION._type,
        id: deviceConfiguration.id,
        name: deviceConfiguration.displayName,
        description:
          deviceConfiguration.description || deviceConfiguration.displayName,
        displayName: deviceConfiguration.displayName,
        version: deviceConfiguration.version,
        configurationType: last(deviceConfiguration['@odata.type']?.split('.')), // Possable values are many, but some examples are iosCustomConfiguration, windows10GeneralConfiguration, iosWiFiConfiguration
      },
    },
  });
}
