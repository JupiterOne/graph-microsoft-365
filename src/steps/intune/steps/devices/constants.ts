import {
  RelationshipClass,
  StepEntityMetadata,
  StepRelationshipMetadata,
} from '@jupiterone/integration-sdk-core';
import { entities as activeDirectoryEntities } from '../../../active-directory';

export const steps: { [k: string]: string } = {
  FETCH_DEVICES: 'managed-devices',
};

export const entities: { [k: string]: StepEntityMetadata } = {
  DEVICE: {
    resourceName: 'Managed Device',
    _type: 'intune_managed_device',
    _class: ['Host', 'Device'],
  },
};

export const relationships: { [k: string]: StepRelationshipMetadata } = {
  USER_HAS_DEVICE: {
    _type: 'microsoft_365_user_has_intune_managed_device',
    sourceType: activeDirectoryEntities.USER._type,
    _class: RelationshipClass.HAS,
    targetType: entities.DEVICE._type,
  },
};
