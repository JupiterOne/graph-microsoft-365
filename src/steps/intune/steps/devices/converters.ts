import {
  createIntegrationEntity,
  createMappedRelationship,
  Entity,
  IntegrationError,
  Relationship,
  RelationshipDirection,
} from '@jupiterone/integration-sdk-core';
import {
  DeviceType,
  ManagedDevice,
} from '@microsoft/microsoft-graph-types-beta';
import { entities as activeDirectoryEntities } from '../../../active-directory';
import { ManagedDeviceType, relationships } from '../../constants';

// https://docs.microsoft.com/en-us/graph/api/resources/intune-devices-manageddevice?view=graph-rest-1.0&viewFallbackFrom=graph-rest-beta
export function createManagedDeviceEntity(
  managedDevice: ManagedDevice,
): Entity {
  const isPhysicalDevice = isPhysical(managedDevice);
  const _class = ['Host'];
  if (isPhysicalDevice) {
    _class.push('Device');
  }
  return createIntegrationEntity({
    entityData: {
      source: managedDevice,
      assign: {
        _class,
        function: ['endpoint-compliance', 'endpoint-protection'], // https://github.com/JupiterOne/data-model/blob/master/src/schemas/HostAgent.json
        _type: selectDeviceType(managedDevice.deviceType, isPhysicalDevice),
        category: 'endpoint',
        id: managedDevice.id,
        name: managedDevice.deviceName,
        deviceName: managedDevice.deviceName,
        displayName: managedDevice.deviceName as string,
        deviceType: managedDevice.deviceType, // desktop, windowsRT, winMO6, nokia, windowsPhone, mac, winCE, winEmbedded, iPhone, iPad, iPod, android, iSocConsumer, unix, macMDM, holoLens, surfaceHub, androidForWork, androidEnterprise, windows10x, androidnGMS, linux, blackberry, palm, unknown, cloudPC.
        model: managedDevice.model,
        make: managedDevice.manufacturer,
        version: managedDevice.model,
        serial: managedDevice.serialNumber,
        serialNumber: managedDevice.serialNumber,
        hardwareVendor:
          managedDevice.hardwareInformation?.manufacturer ??
          managedDevice.manufacturer,
        hardwareManufacturer:
          managedDevice.hardwareInformation?.manufacturer ??
          managedDevice.manufacturer,
        hardwareModel:
          managedDevice.hardwareInformation?.model ?? managedDevice.model,
        hardwareSerial:
          managedDevice.hardwareInformation?.serialNumber ??
          managedDevice.serialNumber,
        hardwareVersion:
          managedDevice.hardwareInformation?.model ?? managedDevice.model,
        meid: managedDevice.meid, // A mobile equipment identifier (MEID) is a globally unique number identifying a physical piece of CDMA2000 mobile station equipment
        imei: managedDevice.imei, // The International Mobile Equipment Identity (IMEI) is a number, usually unique, to identify 3GPP and iDEN mobile phones, as well as some satellite phones.
        iccid: managedDevice.iccid, // Every SIM card has a ICCID number, which stands for Integrated Circuit Card Identifier
        udid: managedDevice.udid, // UDID is an acronym for Unique Device Identifier.
        wifiMacAddress: managedDevice.wiFiMacAddress, // To communicate with a Wi-Fi network, a device must identify itself to the network using a unique network address called a Media Access Control (MAC) address
        easDeviceId: managedDevice.easActivated && managedDevice.easDeviceId, // Microsoft Exchange ActiveSync device identifier
        assetTag:
          managedDevice.azureActiveDirectoryDeviceId ??
          managedDevice.azureADDeviceId, // I believe these are legacy values now
        osDetails: `${managedDevice.operatingSystem} version ${managedDevice.osVersion}`,
        osName: managedDevice.operatingSystem,
        osVersion: managedDevice.osVersion,
        userEmails: managedDevice.emailAddress
          ? [managedDevice.emailAddress]
          : undefined,
        BYOD: managedDevice.managedDeviceOwnerType === 'personal',
        ownerType: managedDevice.managedDeviceOwnerType, // 'personal', 'company' or 'unknown'
        encrypted: managedDevice.isEncrypted,
        managementAgent: managedDevice.managementAgent, // Management channel of the device. Intune, EAS, etc. Possible values are: eas, mdm, easMdm, intuneClient, easIntuneClient, configurationManagerClient, configurationManagerClientMdm, configurationManagerClientMdmEas, unknown, jamf, googleCloudDevicePolicyController
        state: managedDevice.managementState, // "managed", "retirePending", "retireFailed", "wipePending", "wipeFailed", "unhealthy", "deletePending", "retireIssued", "wipeIssued", "wipeCanceled", "retireCanceled", "discovered";
        complianceState: managedDevice.complianceState, // "unknown", "compliant", "noncompliant", "conflict", "error", "inGracePeriod", "configManager";
        compliant:
          managedDevice.complianceState &&
          ['compliant', 'inGracePeriod'].includes(
            managedDevice.complianceState,
          ),
        userId: managedDevice.userId,
        userDisplayName: managedDevice.userDisplayName,
        phoneNumber: managedDevice.phoneNumber,
        managed:
          managedDevice.managementState &&
          managedDevice.managementState !== 'discovered',
        supervised: managedDevice.isSupervised,
        jailBroken: managedDevice.jailBroken !== 'False',
        username: managedDevice.userPrincipalName,
        registrationState: managedDevice.deviceRegistrationState, // Possible values are: notRegistered, registered, revoked, keyConflict, approvalPending, certificateReset, notRegisteredPendingEnrollment, unknown.
        physical: isPhysicalDevice,
        // POTENTIAL: managedDevice.usersLoggedOn - link out to other users perhaps?
      },
    },
  });
}

function isPhysical(managedDevice: ManagedDevice) {
  return managedDevice.model
    ? managedDevice.model !== 'Virtual Machine'
    : undefined; // https://docs.microsoft.com/en-us/mem/intune/fundamentals/windows-10-virtual-machines#reporting
}

/**
 * Select the correct device _type based on the Microsoft DeviceType
 * types determined from managed-questions-endpoint.yaml
 */
export function selectDeviceType(
  deviceType?: DeviceType,
  isPhysical?: Boolean,
): ManagedDeviceType {
  if (!isPhysical) {
    return 'server';
  }
  switch (deviceType) {
    case 'desktop':
    case 'mac':
    case 'unix':
    case 'linux':
      return 'desktop';
    case 'windowsRT': // retired mobile OS
    case 'winMO6': // retired mobile OS
    case 'nokia':
    case 'iPhone':
    case 'iPod':
    case 'android':
    case 'androidForWork': // same OS as android, idk why it is differentiated here
    case 'androidEnterprise': // same OS as android, idk why it is differentiated here
    case 'androidnGMS': // same OS as android, idk why it is differentiated here
    case 'blackberry':
    case 'palm':
      return 'smartphone';
    case 'winCE':
    case 'winEmbedded':
      return 'workstation';
    case 'macMDM':
    case 'windows10x':
      return 'laptop';
    case 'cloudPC':
      return 'server';
    case 'iSocConsumer': // some open source thing
    case 'holoLens': // augmented reality tool
    case 'surfaceHub': // smart whiteboard
    case 'unknown':
      return 'user_endpoint';
    default:
      return 'user_endpoint';
  }
}

/**
 * Creates a mapped relationship between a User and a Device if userId or email is provided.
 * Should be used when the active directory is not being ingested in this integration instance.
 */
export function createUserDeviceMappedRelationship(
  managedDeviceEntity: Entity,
  userId?: string,
  email?: string,
): Relationship | undefined {
  const relationship = relationships.MULTI_USER_HAS_DEVICE.find(
    (relationship) =>
      [relationship.sourceType, relationship.targetType].includes(
        managedDeviceEntity._type,
      ),
  );
  if (!relationship) {
    throw new IntegrationError({
      message: `Could not find relationship for deviceType ${managedDeviceEntity._type}`,
      code: 'RELATIONSHIP_IMPLEMENTATION_ERROR',
    });
  }
  return userId || email
    ? createMappedRelationship({
        _class: relationship._class,
        _type: relationship._type,
        _mapping: {
          sourceEntityKey: managedDeviceEntity._key,
          relationshipDirection: RelationshipDirection.REVERSE,
          targetFilterKeys: [
            ['_class', 'id'],
            ['_class', 'email'],
          ],
          targetEntity: {
            _class: activeDirectoryEntities.USER._class,
            _type: activeDirectoryEntities.USER._type,
            _key: userId,
            id: userId,
            email: email,
          },
        },
      })
    : undefined;
}
