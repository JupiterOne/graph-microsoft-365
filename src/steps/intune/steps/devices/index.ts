import {
  IntegrationStepExecutionContext,
  Step,
} from '@jupiterone/integration-sdk-core';
import { IntegrationConfig, IntegrationStepContext } from '../../../../types';
import { steps as activeDirectorySteps } from '../../../active-directory';
import { DeviceManagementIntuneClient } from '../../clients/deviceManagementIntuneClient';
import { relationships, entities, steps } from './constants';
import {
  createManagedDeviceEntity,
  createUserDeviceDirectRelationship,
  createUserDeviceMappedRelationship,
} from './converters';

export async function fetchDevices(
  executionContext: IntegrationStepContext,
): Promise<void> {
  const { logger, instance, jobState } = executionContext;
  const intuneClient = new DeviceManagementIntuneClient(
    logger,
    instance.config,
  );

  await intuneClient.iterateManagedDevices(async (device) => {
    const deviceEntity = createManagedDeviceEntity(device);
    await jobState.addEntity(deviceEntity);
    const userEntity = await jobState.findEntity(device.userId as string);
    const userDeviceRelationship = userEntity
      ? createUserDeviceDirectRelationship(deviceEntity, userEntity)
      : createUserDeviceMappedRelationship(
          deviceEntity,
          device.userId as string,
          device.emailAddress as string,
        );
    if (userDeviceRelationship) {
      await jobState.addRelationship(userDeviceRelationship);
    }
  });
}

export const deviceSteps: Step<
  IntegrationStepExecutionContext<IntegrationConfig>
>[] = [
  {
    id: steps.FETCH_DEVICES,
    name: 'Managed Devices',
    entities: [entities.DEVICE],
    relationships: [relationships.USER_HAS_DEVICE],
    dependsOn: [activeDirectorySteps.FETCH_USERS],
    executionHandler: fetchDevices,
  },
];
