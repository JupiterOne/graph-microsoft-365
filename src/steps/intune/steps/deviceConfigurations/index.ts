import {
  IntegrationStepExecutionContext,
  Step,
} from '@jupiterone/integration-sdk-core';
import { DeviceConfigurationDeviceOverview } from '@microsoft/microsoft-graph-types-beta';
import { IntegrationConfig, IntegrationStepContext } from '../../../../types';
import { DeviceManagementIntuneClient } from '../../clients/deviceManagementIntuneClient';
import { entities, steps } from '../../constants';
import { createDeviceConfigurationEntity } from './converters';

export async function fetchDeviceConfigurations(
  executionContext: IntegrationStepContext,
): Promise<void> {
  const { logger, instance, jobState } = executionContext;
  const intuneClient = new DeviceManagementIntuneClient(
    logger,
    instance.config,
  );
  await intuneClient.iterateDeviceConfigurations(
    async (deviceConfiguration) => {
      if (!isOrphanedConfig(deviceConfiguration.deviceStatusOverview)) {
        const deviceConfigurationEntity = createDeviceConfigurationEntity(
          deviceConfiguration,
        );
        await jobState.addEntity(deviceConfigurationEntity);
      }
    },
  );
}

export function isOrphanedConfig(
  deviceStatusOverview: DeviceConfigurationDeviceOverview,
) {
  return ![
    'pendingCount',
    'notApplicableCount',
    'notApplicablePlatformCount',
    'successCount',
    'errorCount',
    'failedCount',
    'conflictCount',
  ].find((countKey) => deviceStatusOverview[countKey]);
}

export const deviceConfigurationSteps: Step<
  IntegrationStepExecutionContext<IntegrationConfig>
>[] = [
  {
    id: steps.FETCH_DEVICE_CONFIGURATIONS,
    name: 'Device Configurations',
    entities: [entities.DEVICE_CONFIGURATION],
    relationships: [], // (Device => Device Configuration) relationship is created in the FETCH_NONCOMPLIANCE_FINDINGS step
    dependsOn: [],
    executionHandler: fetchDeviceConfigurations,
  },
];
