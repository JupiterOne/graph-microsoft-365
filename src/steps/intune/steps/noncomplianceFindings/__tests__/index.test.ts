import {
  createMockStepExecutionContext,
  Recording,
} from '@jupiterone/integration-sdk-testing';
import { setupAzureRecording } from '../../../../../../test/recording';
import { config } from '../../../../../../test/config';
import { fetchNonComplianceFindings } from '..';
import { fetchDevices } from '../../devices';
import { fetchDeviceConfigurations } from '../../deviceConfigurations';
import { entities, relationships } from '../../../constants';
import { Entity } from '@jupiterone/integration-sdk-core';
import { isEqual } from 'lodash';
import { ensureArray } from '../../../../../../test/ensureArray';

let recording: Recording;

afterEach(async () => {
  if (recording) {
    await recording.stop();
  }
});

describe('fetchNonComplianceFindings', () => {
  test('should make entities and relationships correctly', async () => {
    recording = setupAzureRecording({
      directory: __dirname,
      name: 'fetchNonComplianceFindings',
    });
    const context = createMockStepExecutionContext({ instanceConfig: config });

    await fetchDevices(context);
    await fetchDeviceConfigurations(context);
    await fetchNonComplianceFindings(context);

    const noncomplianceFindingEntities = context.jobState.collectedEntities.filter(
      (e) =>
        isEqual(e._class, ensureArray(entities.NONCOMPLIANCE_FINDING._class)),
    );
    const noncomplianceFindingDeviceRelationships = context.jobState.collectedRelationships.filter(
      (r) =>
        isEqual(r._type, relationships.DEVICE_USES_DEVICE_CONFIGURATION._type),
    );
    const noncomplianceFindingDeviceConfigurationRelationships = context.jobState.collectedRelationships.filter(
      (r) =>
        isEqual(
          r._type,
          relationships.DEVICE_CONFIGURATION_IDENTIFIED_NONCOMPLIANCE_FINDING
            ._type,
        ),
    );
    const deviceDeviceConfigurationRelationships = context.jobState.collectedRelationships.filter(
      (r) =>
        isEqual(r._type, relationships.DEVICE_HAS_NONCOMPLIANCE_FINDING._type),
    );

    // Check that we have Noncompliance Findings
    expect(noncomplianceFindingEntities.length).toBeGreaterThan(0);
    expect(noncomplianceFindingEntities).toMatchGraphObjectSchema({
      _class: entities.NONCOMPLIANCE_FINDING._class,
    });
    expect(noncomplianceFindingEntities).toMatchSnapshot();
    // Check that only open findings are created
    noncomplianceFindingEntities.forEach((entity: Entity) => {
      expect(entity.open).toBe(true);
    });

    // Check that we have DEVICE_USES_DEVICE_CONFIGURATION relationships
    expect(noncomplianceFindingDeviceRelationships.length).toBeGreaterThan(0);
    expect(
      noncomplianceFindingDeviceRelationships,
    ).toMatchDirectRelationshipSchema({});
    expect(noncomplianceFindingDeviceRelationships).toMatchSnapshot(
      'noncomplianceFindingDeviceRelationships',
    );

    // Check that we have DEVICE_CONFIGURATION_IDENTIFIED_NONCOMPLIANCE_FINDING relationships
    expect(
      noncomplianceFindingDeviceConfigurationRelationships.length,
    ).toBeGreaterThan(0);
    expect(
      noncomplianceFindingDeviceConfigurationRelationships,
    ).toMatchDirectRelationshipSchema({});
    expect(
      noncomplianceFindingDeviceConfigurationRelationships,
    ).toMatchSnapshot('noncomplianceFindingDeviceConfigurationRelationships');

    // Check that we have DEVICE_HAS_NONCOMPLIANCE_FINDING relationships
    expect(deviceDeviceConfigurationRelationships.length).toBeGreaterThan(0);
    expect(
      deviceDeviceConfigurationRelationships,
    ).toMatchDirectRelationshipSchema({});
    expect(deviceDeviceConfigurationRelationships).toMatchSnapshot(
      'deviceDeviceConfigurationRelationships',
    );
  });
});
