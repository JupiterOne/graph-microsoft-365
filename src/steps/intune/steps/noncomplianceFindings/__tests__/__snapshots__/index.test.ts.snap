// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`fetchNonComplianceFindings should make entities and relationships correctly 1`] = `
Array [
  Object {
    "_class": Array [
      "Finding",
    ],
    "_key": "7b521133-f056-40a9-b639-b76afd67e4bb_6aa605f0-744e-4ae1-8031-c6564624c89b_683dbff1-c8ff-4996-91ae-85484de46cfc",
    "_rawData": Array [
      Object {
        "name": "default",
        "rawData": Object {
          "complianceGracePeriodExpirationDateTime": "2021-02-17T19:13:41.8113Z",
          "deviceDisplayName": "[REDACTED]",
          "deviceModel": null,
          "id": "7b521133-f056-40a9-b639-b76afd67e4bb_6aa605f0-744e-4ae1-8031-c6564624c89b_683dbff1-c8ff-4996-91ae-85484de46cfc",
          "lastReportedDateTime": "2021-03-01T16:23:39.1872164Z",
          "status": "error",
          "userName": "[REDACTED]",
          "userPrincipalName": "[REDACTED]",
        },
      },
    ],
    "_type": "intune_noncompliance_finding",
    "active": undefined,
    "assessment": "Test configuration policy",
    "category": "endpoint",
    "createdOn": undefined,
    "displayName": "Latest finding from Test configuration policy",
    "id": "7b521133-f056-40a9-b639-b76afd67e4bb_6aa605f0-744e-4ae1-8031-c6564624c89b_683dbff1-c8ff-4996-91ae-85484de46cfc",
    "lastProcessedOn": 1614615819187,
    "lastTestedOn": 1614615819187,
    "lastUpdatedOn": 1614615819187,
    "name": "Latest finding from Test configuration policy",
    "numericSeverity": 6,
    "open": true,
    "severity": "medium",
    "status": "error",
  },
]
`;

exports[`fetchNonComplianceFindings should make entities and relationships correctly: deviceDeviceConfigurationRelationships 1`] = `
Array [
  Object {
    "_class": "HAS",
    "_fromEntityKey": "683dbff1-c8ff-4996-91ae-85484de46cfc",
    "_key": "683dbff1-c8ff-4996-91ae-85484de46cfc|has|7b521133-f056-40a9-b639-b76afd67e4bb_6aa605f0-744e-4ae1-8031-c6564624c89b_683dbff1-c8ff-4996-91ae-85484de46cfc",
    "_toEntityKey": "7b521133-f056-40a9-b639-b76afd67e4bb_6aa605f0-744e-4ae1-8031-c6564624c89b_683dbff1-c8ff-4996-91ae-85484de46cfc",
    "_type": "intune_managed_device_has_noncompliance_finding",
    "displayName": "HAS",
  },
]
`;

exports[`fetchNonComplianceFindings should make entities and relationships correctly: noncomplianceFindingDeviceConfigurationRelationships 1`] = `
Array [
  Object {
    "_class": "IDENTIFIED",
    "_fromEntityKey": "6aa605f0-744e-4ae1-8031-c6564624c89b",
    "_key": "6aa605f0-744e-4ae1-8031-c6564624c89b|identified|7b521133-f056-40a9-b639-b76afd67e4bb_6aa605f0-744e-4ae1-8031-c6564624c89b_683dbff1-c8ff-4996-91ae-85484de46cfc",
    "_toEntityKey": "7b521133-f056-40a9-b639-b76afd67e4bb_6aa605f0-744e-4ae1-8031-c6564624c89b_683dbff1-c8ff-4996-91ae-85484de46cfc",
    "_type": "intune_device_configuration_identified_noncompliance_finding",
    "displayName": "IDENTIFIED",
  },
]
`;

exports[`fetchNonComplianceFindings should make entities and relationships correctly: noncomplianceFindingDeviceRelationships 1`] = `
Array [
  Object {
    "_class": "USES",
    "_fromEntityKey": "683dbff1-c8ff-4996-91ae-85484de46cfc",
    "_key": "683dbff1-c8ff-4996-91ae-85484de46cfc|uses|6aa605f0-744e-4ae1-8031-c6564624c89b",
    "_toEntityKey": "6aa605f0-744e-4ae1-8031-c6564624c89b",
    "_type": "intune_managed_device_uses_device_configuration",
    "displayName": "USES",
  },
]
`;
